rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,783 @@
+            saveNodeAndCreateItem(newNode, commPrefix, optimization = false).some.pure
+          }
+      }
+
+    /**
+      * Delete leaf value from this part of tree (start from curItem).
+      * Rehash and save all depend node to cacheW. Return updated curItem.
+      * If not found leaf with  delPrefix - return [[None]].
+      */
+    def delete(curItem: Item, delPrefix: ByteVector): F[Option[Item]] =
+      curItem match {
+        case EmptyItem => none[Item].pure //Not found
+
+        case Leaf(leafPrefix, _) =>
+          if (leafPrefix == delPrefix) (EmptyItem: Item).some.pure //Happy end
+          else none[Item].pure                                     //Not found
+
+        case NodePtr(ptrPrefix, ptr) =>
+          val (commPrefix, delPrefixRest, ptrPrefixRest) = commonPrefix(delPrefix, ptrPrefix)
+          if (ptrPrefixRest.nonEmpty || delPrefixRest.isEmpty) none[Item].pure //Not found
+          else {
+            val (childItemIdx, childDelPrefix) =
+              (byteToInt(delPrefixRest.head), delPrefixRest.tail)
+            for {
+              childNode    <- loadNode(ptr)
+              childItemOpt <- delete(childNode(childItemIdx), childDelPrefix) //deeper
+              returnedChild = childItemOpt match {
+                case None => none
+                case Some(childItem) =>
+                  saveNodeAndCreateItem(childNode.updated(childItemIdx, childItem), commPrefix).some
+              }
+            } yield returnedChild
+          }
+      }
+
+    /**
+      * Parallel processing of HistoryActions in this part of tree (start from curNode).
+      * New data load to [[cacheW]].
+      * Return updated curNode. if no action was taken - return [[None]].
+      */
+    def makeActions(curNode: Node, curActions: List[HistoryAction]): F[Option[Node]] = {
+      val groups = curActions
+        .groupBy( //fixme
+          _.key.headOption.getOrElse({
+            assert(assertion = false, "The length of all prefixes in the subtree must be the same")
+            0x00.toByte
+          })
+        )
+        .toList
+      val newGroupItemsF = groups.map {
+        case (groupIdx, groupActions) =>
+          val index = byteToInt(groupIdx)
+          val item  = curNode(index)
+          if (groupActions.length == 1) {
+            val newItem = groupActions.head match {
+              case InsertAction(key, hash) =>
+                update(item, ByteVector(key).tail, hash.bytes)
+              case DeleteAction(key) => delete(item, ByteVector(key).tail)
+            }
+            newItem.map((index, _))
+          } else {
+            val notExistInsertAction = groupActions.collectFirst { case _: InsertAction => true }.isEmpty
+            val clearGroupActions =
+              if (item == EmptyItem && notExistInsertAction) groupActions.collect {
+                case v: InsertAction => v
+              } else groupActions
+            if (clearGroupActions.isEmpty) (index, none[Item]).pure
```

Comment:
Just a question. line 734 to 739 are just some optimizations on some simple situations, aren't they? I don't actually understand what `clearGroupActions`  means.

Reply:
if we don't have data in this branch (in this we have EmptyItem) and if we don't wanna insert data in this branch, we don't create the parallel process for  DeleteAction in this branch.